builtin/credential/aws/path_role.go

```diff
@@ -203,7 +203,49 @@ func (b *backend) lockedAWSRole(s logical.Storage, roleName string) (*awsRoleEnt
 	b.roleMutex.RLock()
 	defer b.roleMutex.RUnlock()
 
-	return b.nonLockedAWSRole(s, roleName)
+	roleEntry, err := b.nonLockedAWSRole(s, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, nil
+	}
+	needUpgrade, err := upgradeRoleEntry(roleEntry)
+	if err != nil {
+		return nil, fmt.Errorf("error upgrading roleEntry: %v", err)
+	}
+	if needUpgrade {
+		// we need to get a write lock to upgrade the role entry, so we first need to release our read lock
+		// to prevent the write lock from deadlocking on it
+		b.roleMutex.RUnlock()
+		// now we need to ensure that we re-acquire the read lock so the above deferred RUnlock() doesn't
+		// cause a crash trying to unlock the already-unlocked mutex
+		defer b.roleMutex.RLock()
```

Comment:
Oops, sorry, missed that case. I had thought about it and forgot to mention while typing up the comment. We need to explicitly release the locks and return. If `needUpgrade` is `true` we release the write lock and return, if not we release the read lock and return. Probably `defer` statements just before the return will also do.